app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -0,0 +1,248 @@
+import { ethErrors } from 'eth-rpc-errors';
+import validUrl from 'valid-url';
+import { omit } from 'lodash';
+import { MESSAGE_TYPE } from '../../../../../shared/constants/app';
+import {
+  isPrefixedFormattedHexString,
+  isSafeChainId,
+} from '../../../../../shared/modules/network.utils';
+import { jsonRpcRequest } from '../../../../../shared/modules/rpc.utils';
+import { CHAIN_ID_TO_NETWORK_ID_MAP } from '../../../../../shared/constants/network';
+
+const addEthereumChain = {
+  methodNames: [MESSAGE_TYPE.ADD_ETHEREUM_CHAIN],
+  implementation: addEthereumChainHandler,
+};
+export default addEthereumChain;
+
+async function addEthereumChainHandler(
+  req,
+  res,
+  _next,
+  end,
+  {
+    addCustomRpc,
+    getCurrentNetwork,
+    findCustomRpcBy,
+    updateRpcTarget,
+    requestUserApproval,
+  },
+) {
+  if (!req.params?.[0] || typeof req.params[0] !== 'object') {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected single, object parameter. Received:\n${req.params}`,
+      }),
+    );
+  }
+
+  const { origin } = req;
+
+  const {
+    chainId,
+    chainName = null,
+    blockExplorerUrls = null,
+    nativeCurrency = null,
+    rpcUrls,
+  } = req.params[0];
+
+  const otherKeys = Object.keys(
+    omit(req.params[0], [
+      'chainId',
+      'chainName',
+      'blockExplorerUrls',
+      'iconUrls',
```

Comment:
I notice that `'iconUrls'` is not included in `otherKeys`, but neither is it used anywhere. Is there some reason we are allowing this in the request params even though it will go unused?

Reply:
It is a part of EIP-3085, the goal here being to warn for extraneous keys that are not a part of the EIP. For example: 'blockExplorerUrls' is what we want, but someone might put 'blockExplorerUrl'. Beings as blockExplorerUrls is optional, without this check we'd accept the request and add the chain to the user's MetaMask without the blockExplorer, and the dapp would be none the wiser.